### PR TITLE
AzCopy sync with --hash-meta-dir causes original files to get hidden

### PR DESCRIPTION
**Pull request by dphulkar-msft**
**Command:**
.\azure-storage-azcopy.exe sync "C:\testdata" "https://dpazcopy2.blob.core.windows.net/samplecont?<sas-token>" --delete-destination true --exclude-path "backup;import;system/caches" --compare-hash "MD5" --put-md5 --cap-mbps 200 --local-hash-storage-mode HiddenFiles --hash-meta-dir "C:\hash-meta"

**Issue:** https://github.com/Azure/azure-storage-azcopy/issues/2758

**BugFix:**
       According to the command the metadata file should be hidden and not the original file.
SS2

image
image

### CODE REVIEW

**Review by adreed-msft — APPROVED** [review]

**Comment by adreed-msft** [comment]
We should add an automated test to this, but the change itself looks good.

**Comment by seanmcc-msft** [comment]
> We should add an automated test to this, but the change itself looks good.

Yes, please add a test case.

**Review by adreed-msft — CHANGES_REQUESTED** [review]
rc: test case needed

**Comment by dphulkar-msft** [comment]
> rc: test case needed

**Comment by dphulkar-msft** [comment]
@adreed-msft  @vibhansa-msft Added the testcase. Please review

File: e2etest/zt_basic_copy_sync_remove_test.go [hunk 1]
```diff
@@ -860,6 +862,9 @@ func TestBasic_HashBasedSync_HashDir(t *testing.T) {
 				hashFile := filepath.Join(hashStorageDir, ".asdf.txt"+common.AzCopyHashDataStream)
 				_, err = os.Stat(hashFile)
 				a.AssertNoErr(err)
+
+				isHidden := osScenarioHelper{}.IsFileHidden(a, hashFile)
```
Comment (adreed-msft) [inline e2etest/zt_basic_copy_sync_remove_test.go:866]: Quick check: make sure the data file is also not hidden. Otherwise, lgtm.

Comment (vibhansa-msft) [reply e2etest/zt_basic_copy_sync_remove_test.go:866]: Yes we shall check both ways, meta shall be hidden and data shall not be hidden (which is the exact issue we are trying to solve here)

File: common/hash_data_adapter_hidden_files.go [hunk 2]
```diff
@@ -66,9 +67,18 @@ func (a *HiddenFileDataAdapter) SetHashData(relativePath string, data *SyncHashD
...
+	_, err := os.Stat(metaFile)
+	if os.IsNotExist(err) {
+		f, err = os.OpenFile(metaFile, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to create hash meta file: %w", err)
+		}
+	} else {
+		f, err = os.OpenFile(metaFile, os.O_TRUNC|os.O_RDWR, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to open hash meta file: %w", err)
```
Comment (vibhansa-msft) [inline common/hash_data_adapter_hidden_files.go:80]: error log going to console, rather shall go to log file.

File: common/hash_data_adapter_hidden_files.go [hunk 3]
```diff
@@ -66,9 +67,18 @@ func (a *HiddenFileDataAdapter) SetHashData(relativePath string, data *SyncHashD
 
 	metaFile := a.getHashPath(relativePath)
 
-	f, err := os.OpenFile(metaFile, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0644)
-	if err != nil {
-		return fmt.Errorf("failed to open hash meta file: %w", err)
+	var f *os.File
+	_, err := os.Stat(metaFile)
+	if os.IsNotExist(err) {
```
Comment (vibhansa-msft) [inline common/hash_data_adapter_hidden_files.go:72]: Both if and else are doing almost the same thing, we shall rather unify the case. Only the log differs it seems here.

File: common/hash_data_adapter_hidden_files.go [hunk 4]
```diff
@@ -83,9 +93,8 @@ func (a *HiddenFileDataAdapter) SetHashData(relativePath string, data *SyncHashD
 
 	// Push types around to check for OS-specific hide file method
 	if adapter, canHide := any(a).(interface{ HideFile(string) error }); canHide {
-		dataFile := a.getDataPath(relativePath)
-
-		err := adapter.HideFile(dataFile)
+		metaFile := a.getHashPath(relativePath)
```
Comment (vibhansa-msft) [inline common/hash_data_adapter_hidden_files.go:96]: We shall put some comments here on why we are changing this code and why meta shall be hidden

File: common/hash_data_adapter_hidden_files.go (hunk 3 above)
Comment (dphulkar-msft) [reply common/hash_data_adapter_hidden_files.go:72]: There is one use case where it fails if we use a single OpenFile() API with os.O_CREATE|os.O_TRUNC|os.O_RDWR. 
Once the metafile is hidden and if we try to run the same azcopy command it fails to open the meta file with these flags being used combined os.O_CREATE|os.O_TRUNC|os.O_RDWR for which I have seperated the 2 calls.

File: common/hash_data_adapter_hidden_files.go (hunk 4 above)
Comment (dphulkar-msft) [reply common/hash_data_adapter_hidden_files.go:96]: added

File: e2etest/zt_basic_copy_sync_remove_test.go (hunk 1 above)
Comment (dphulkar-msft) [reply e2etest/zt_basic_copy_sync_remove_test.go:866]: added

**Review by vibhansa-msft — APPROVED** [review]

File: common/hash_data_adapter_hidden_files.go (hunk 3 above)
Comment (dphulkar-msft) [reply common/hash_data_adapter_hidden_files.go:74]: Added the comment in the code as well.

**Pull request merged**